torch/utils/data/_utils/collate.py

```diff
@@ -26,11 +26,19 @@ def default_convert(data):
             return data
         return torch.as_tensor(data)
     elif isinstance(data, collections.abc.Mapping):
-        return {key: default_convert(data[key]) for key in data}
+        try:
+            return elem_type({key: default_convert(data[key]) for key in data})
+        except TypeError:
+            # The mapping type may not support `__init__(iterable)`.
+            return {key: default_convert(data[key]) for key in data}
     elif isinstance(data, tuple) and hasattr(data, '_fields'):  # namedtuple
         return elem_type(*(default_convert(d) for d in data))
     elif isinstance(data, collections.abc.Sequence) and not isinstance(data, string_classes):
-        return [default_convert(d) for d in data]
+        try:
+            return elem_type([default_convert(d) for d in data])
+        except TypeError:
+            # The sequence type may not support `__init__(iterable)` (e.g., `range`).
+            return [default_convert(d) for d in data]
```

Comment:
This is part is BC-breaking for the case that `elem_type` is tuple. Before this PR, we converts tuple to list. But now, it remains as tuple.

To preserve backward compatible, could you please also add conversion from tuple to list if `elem_type` is tuple.

I need to discuss with some users who are relying on such behavior and see if we can allow such BC-breaking change.